Feature/gfsv16b paranetcdf

yangfanglin opened the pull request:
While I am still testing this branch with parallel netCDF enabled, I'd like to give you the opportunity to review the changes  before the testing is completed.   The test was set up to run one and an half cycle.  Changes have already been made to the external repos, including GSI, UPP, GLDAS and UFS_UTIL, to compile with the new parallel netcdf libs and associated esmf lib.   Code managers are yet to make new tags/branches available.  I am still checking if there is any other external apps need to be updated.

yangfanglin committed:
The following files are no longer used
 	deleted:    README.iau
 	deleted:    config.anal.iau
 	deleted:    config.base.emc.dyn.iau
 	deleted:    config.efcs.iau
 	deleted:    config.eobs.iau
 	deleted:    config.eupd.iau
 	deleted:    config.vrfy.iau

yangfanglin committed:
Github Issue #16

1. add options to workflow scripts to use parallel netcdf for I/O.
   The application varies with model resolution and computing platform.
   See config.fcst and exglobal_fcst_nemsfv3gfs.sh for the settings.

2. Update UPP to upp_gfsv16_release.v1.0.1

 	modified:   modulefiles/module_base.hera
 	modified:   modulefiles/module_base.wcoss_dell_p3
 	modified:   parm/config/config.fcst
 	modified:   scripts/exglobal_fcst_nemsfv3gfs.sh
 	modified:   sorc/checkout.sh
 	deleted:    parm/config/config.resources.C96

yangfanglin committed:
Modified config.resources to use npe_node_max to define computing resouces. Pull over new settings found in
the config files of in v16rt2.  Made ertain adjustment and cleaned up certain parameters in a few scripts to
remove redundance.

 	modified:   parm/config/config.anal
 	modified:   parm/config/config.base.emc.dyn
 	modified:   parm/config/config.efcs
 	modified:   parm/config/config.eobs
 	modified:   parm/config/config.eupd
 	modified:   parm/config/config.fcst
 	modified:   parm/config/config.fv3
 	modified:   parm/config/config.gldas
 	modified:   parm/config/config.post
 	modified:   parm/config/config.resources

yangfanglin committed:
modified:   parm/config/config.fcst
 	modified:   parm/config/config.fv3

RussTreadon-NOAA committed:
Issue #16:  update DA checkout to feature/parallel_ncio

yangfanglin committed:
modified:   modulefiles/module_base.hera and   modulefiles/module_base.wcoss_dell_p3 to point to parallel versions of netCDF libs and esmf lib
 	modified:   parm/config/config.efcs and  parm/config/config.fcst to set chunksizes correctly for high-res and enkf forecasts

yangfanglin committed:
modified:   config.resources

yangfanglin commented:
The changes are rather significant, including
1) running forecasts with parallel netCDF read and write, with different options for different model resolutions and on different platforms.
2) adding all options currently used by v16rt2 into the workflow
3) removing those "*.iau" configure files
4) restructuring config.fv3 and config.resources so that the computation resources can be optimally set  for different computers based on the max cores per node setting.  Removed many redundant and sometimes conflicting parameters. 

Supposedly, out of the box the branch can reproduce v16rt2.

yangfanglin committed:
modified:   parm/config/config.base.emc.dyn to point new obsproc_prep to spupport parallel netcdf

RussTreadon-NOAA committed:
Issue #16:  Update EMC_verif-global to tag verif_global_v1.5.0

CoryMartin-NOAA commented on modulefiles/module_base.hera:
```diff
@@ -7,22 +7,27 @@ module load intel/18.0.5.274
-module load hdf5/1.10.5
 module load gempak/7.4.2
 
+#module load pnetcdf/1.10.0
```
Do we want to keep the commented out modules? I guess it's okay for now

yangfanglin replied:
removed

CoryMartin-NOAA commented on parm/config/config.eupd:
```diff
@@ -11,6 +11,25 @@ echo "BEGIN: config.eupd"
+export letkf_flag=".true."       # use LETKF instead of serial filter
+export getkf=".true."            # Gain form of LETKF (needed for model-space localization)
+export denkf=".true."            # EnKF approximation (beneficial since less spread removed by analysis)
+export nobsl_max=10000           # max number of obs in each LETKF volume (uses closest nobsl_max). can
```
are we decreasing this because of runtime concerns?

yangfanglin replied:
Cory:  these parameters are copied over from v16rt2.

CoryMartin-NOAA replied:
@yangfanglin yes sorry, I meant more for asking @RussTreadon-NOAA if we decided if we need to decrease from 10000 due to runtime concerns from testing Cathy has done.

RussTreadon-NOAA replied:
Good question.  Let's ask Cathy what the DA team decided.  I don't recall.

CoryMartin-NOAA commented on parm/config/config.gldas:
```diff
@@ -12,9 +12,7 @@ export GLDASSH=$HOMEgfs/scripts/exgdas_gldas.sh.ecf
 
 
-if [ $machine = "WCOSS_C" -o $machine = "WCOSS_DELL_P3" ]; then
-    export CPCGAUGE=/gpfs/dell2/emc/modeling/noscrub/Fanglin.Yang/stat/cpcUniGauge
```
Is this supposed to be removed for Cray/Dell?

yangfanglin replied:
already removed see the right hand column

CoryMartin-NOAA commented on parm/config/config.resources:
```diff
@@ -40,87 +40,52 @@ if [ $step = "prep" -o $step = "prepbufr" ]; then
 elif [ $step = "anal" ]; then
 
-    export wtime_anal="02:30:00"
+    export wtime_anal="06:00:00"
```
Why has this been bumped up to 6 hours? Won't this cause the job to queue longer than necessary?

KateFriedman-NOAA replied:
I have the same question as Cory.

yangfanglin replied:
Reduced to 2 hours.

CoryMartin-NOAA commented on parm/config/config.resources:
```diff
@@ -196,61 +160,40 @@ elif [ $step = "eupd" ]; then
 elif [ $step = "ecen" ]; then
 
-    export wtime_ecen="00:30:00"
+    export wtime_ecen="03:00:00"
```
Again, I guess the general comment for a lot of these changes, is why are the wall time requests so much larger, this one is 6x as long. Is that actually what is needed now?

yangfanglin replied:
changed back to 30 minutes.
Most of  the numbers are copied over from v16rt2, in which some of the larger numbers were set during trouble shootings and testing.  All these numbers need to be revisited before the gfsv16b branches are merged back to develop.

RussTreadon-NOAA replied:
v16rt2 wall times are inflated to minimize failures due to long running jobs.    All GFS v16 resource and wall clock limits need to be revisited.   v16rt2 has been and remains a moving target.  It is not a mature, ready-to-implement package.

KateFriedman-NOAA commented on parm/config/config.anal:
```diff
@@ -93,9 +104,17 @@ if [ $REALTIME = "NO" ]; then
-if [ $LEVS = "128" ]; then
-   export GRIDOPTS="nlayers(63)=1,nlayers(64)=1,"
-   export SETUP="gpstop=55,nsig_ext=56,$SETUP"
+# TEST - use experimental dumps
```
Will need to remember to remove pointers to x-dumps or wrap in a date if-block so the experimental versions are used in the future for dates before these files went into production.

yangfanglin replied:
good point.  This is copied over from v16rt2. Need to check when and for how long these two have been used in the parallel.   For now, my purpose if to make sure this branch can reproduce v16rt2

RussTreadon-NOAA replied:
Please contact Haixia to get start dates for AHI and ABI.

KateFriedman-NOAA commented on parm/config/config.fcst:
```diff
@@ -138,6 +138,31 @@ export ishuffle=0
+            export OUTPUT_FILETYPES=" 'netcdf_parallel' 'netcdf' "
+        fi
+    fi
+    if [[ "$machine" == "HERA" ]]; then
```
Why is Hera different here? Mostly just curious.

yangfanglin replied:
This setting is based on extensive tests made by Jun Wang and Jeff Whitaker.  It is odd, but indeed writing the 2D files using parallel_netcdf is slow than serial netcdf on WCCOSS but faster on HERA.

KateFriedman-NOAA replied:
Got it, thanks!

KateFriedman-NOAA commented on parm/config/config.gldas:
```diff
@@ -12,9 +12,7 @@ export GLDASSH=$HOMEgfs/scripts/exgdas_gldas.sh.ecf
+if [ $machine = "HERA" ]; then
     export CPCGAUGE=/scratch1/NCEPDEV/global/Fanglin.Yang/stat/cpcUniGauge
 fi
 
```
The CPC gauge files are in the Hera GDA just like in the WCOSS GDA so this isn't necessary anymore.

yangfanglin replied:
already removed in the new branch

KateFriedman-NOAA commented on parm/config/config.resources:
```diff
@@ -40,87 +40,52 @@ if [ $step = "prep" -o $step = "prepbufr" ]; then
-    fi
     export wtime_fcst="01:00:00"
-    export wtime_fcst_gfs="06:00:00"
+    export wtime_fcst_gfs="10:00:00"
```
Hera's max walltime is 8hrs so this won't work on Hera. Perhaps we need to introduce a wtime_max variable for each machine and use it here?

yangfanglin replied:
reduced to 8 hrs.  Should use much less, but this is to allow the fcst to finish if low resources are given.

KateFriedman-NOAA commented on parm/config/config.resources:
```diff
@@ -159,7 +124,7 @@ elif [ $step = "eobs" -o $step = "eomg" ]; then
-      export npe_eobs=84
+      export npe_eobs=400
     elif [ $CASE = "C384" ]; then
       export npe_eobs=42
```
Do the lower resolutions need a similar increase for npe_eobs? Going from 400 for C768 to 42 for C384 seems too big of a drop now.

yangfanglin replied:
Need to test and make changes in the future.

RussTreadon-NOAA reviewed (commented):
We need the changes in this branch for GFS v16.   If fv3test reproduces v16rt2, merge.  If fv3test does not reproduce v16rt2, identify steps which differ.  If difference is expected and acceptable, merge.   

GFS v16 components need to be tagged before official GFSv16 parallel begin.    DA tag can not yet be created since some items affecting results remain under active investigation.   This does not prevent this branch from being merged.

yangfanglin committed:
Point new UPP tag upp_gfsv16_release.v1.0.2, for upgrading with netcdf_parallel 4.7.4, hdf_parallel 1.10.6 and
w3emc_para 2.4.0 on Dell and Hera

yangfanglin committed:
Update to use parallel netcdf libs
 	modified:   modulefiles/fv3gfs/enkf_chgres_recenter.wcoss_dell_p3
 	modified:   modulefiles/fv3gfs/enkf_chgres_recenter_nc.hera
 	modified:   modulefiles/fv3gfs/enkf_chgres_recenter_nc.wcoss_dell_p3
 	modified:   modulefiles/fv3gfs/gaussian_sfcanl.hera
 	modified:   modulefiles/fv3gfs/gaussian_sfcanl.wcoss_dell_p3

yangfanglin committed:
modified:   ../../modulefiles/module_base.hera
 	modified:   config.anal        control ABI and AHI data usage
 	modified:   config.resources   updates based on pull reviews
 	modified:   ../../sorc/gaussian_sfcanl.fd/makefile.sh  use parallel netcdf

yangfanglin committed:
modified:   sorc/checkout.sh update ufs_util repo

CoryMartin-NOAA committed:
Added namelist option 'cld_amt' to enkf_chgres_recenter_ncio. Default is .false., when .true. will attempt to read in cld_amt and print error (but continue) if it is missing from input file

yangfanglin committed:
modified:   parm/config/config.base.emc.dyn
 	modified:   parm/config/config.base.nco.static

yangfanglin committed:
Merge branch 'feature/gfsv16b_paranetcdf' of https://github.com/NOAA-EMC/global-workflow into feature/gfsv16b_paranetcdf

yangfanglin commented:
The experiment fv3test1 has been run successfully for all steps of the gdas cycle. Ready to merge this branch gfsv16v_paranetcdf back to gfsv16b
